xls/modules/rle/rle.x

```diff
@@ -0,0 +1,322 @@
+// last                                   ┌─────┐           ┌─────┐
+// (input channel)      ──────────────────┘     └───────────┘     └──────────
+//                      ╥─────╥─────╥─────╥─────╥─────╥─────╥─────╥──────────
+// state.prev_symbol    ║ 0   ║ A   ║ A   ║ B   ║ 0   ║ B   ║ C   ║ 0
+// (set state value)    ╨─────╨─────╨─────╨─────╨─────╨─────╨─────╨──────────
+//                      ╥─────╥─────╥─────╥─────╥─────╥─────╥─────╥──────────
+// state.prev_count     ║ 0   ║ 1   ║ 2   ║ 1   ║ 0   ║ 1   ║ 1   ║ 0
+// (set state value)    ╨─────╨─────╨─────╨─────╨─────╨─────╨─────╨──────────
+//
+// do_send                                ┌───────────┐     ┌───────────┐
+//                      ──────────────────┘           └─────┘           └────
+//                      ──────────────────╥─────╥─────╥─────╥─────╥─────╥────
+// symbol, count        XXXXXXXXXXXXXXXXXX║ A,2 ║ B,1 ║XXXXX║ B,1 ║ C,1 ║XXXX
+// (output channel)     ──────────────────╨─────╨─────╨─────╨─────╨─────╨────
+// last                                         ┌─────┐           ┌─────┐
+// (output channel)     ────────────────────────┘     └───────────┘     └────
+
+import std
+
+// structure containing input data for an RLE encoder
+pub struct EncInData<SYMBOL_WIDTH: u32> {
+    symbol: bits[SYMBOL_WIDTH], // input symbol to compress
+    last: bool,                 // if this is the last symbol
+}
+
+// structure containing output data from an RLE encoder
+pub struct EncOutData<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    symbol: bits[SYMBOL_WIDTH], // compressed symbol
+    count: bits[COUNT_WIDTH],   // symbol count
+    last: bool,                 // if this is the last symbol
+}
+
+// structure to preserve the state of an RLE encoder
+pub struct RLEEncState<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    // symbol from the previous RLEEnc::next evaluation,
+    // valid if prev_count > 0
+    prev_symbol: bits[SYMBOL_WIDTH],
+    // symbol count from the previous RLEEnc::next evaluation.
+    // zero means that the previous evaluation sent all the data and
+    // we start counting from the beginning
+    prev_count: bits[COUNT_WIDTH],
+    // flag indicating that the previous symbol was the last one
+    // in the transmission
+    prev_last: bool,
+}
+
+// RLE encoder implementation
+pub proc RLEEnc<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    input_r: chan<EncInData<SYMBOL_WIDTH>> in;
+    output_s: chan<EncOutData<SYMBOL_WIDTH, COUNT_WIDTH>> out;
+
+    init {(
+        RLEEncState<SYMBOL_WIDTH, COUNT_WIDTH> {
+            prev_symbol: bits[SYMBOL_WIDTH]:0,
+            prev_count: bits[COUNT_WIDTH]:0,
+            prev_last: false,
+        }
+    )}
+
+    config (
+        input_r: chan<EncInData<SYMBOL_WIDTH>> in,
+        output_s: chan<EncOutData<SYMBOL_WIDTH, COUNT_WIDTH>> out,
+    ) {(input_r, output_s)}
+
+    next (tok: token, state: RLEEncState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+        let zero_input = EncInData { symbol: bits[SYMBOL_WIDTH]:0, last: false };
```

Comment:
nit: wonder if we could move this as a top level `const`, wdyt?

Reply:
This is a parametric struct, I tried to set `zero_input` as a parametric value of `RLEEnc`, but it seems that a parametric value cannot be of parametric struct type. I may open a separate issue for that, unless it is an expected behavior.